click_odoo_contrib/backupdb.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright 2018 ACSONE SA/NV (<http://acsone.eu>)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+import json
+import os
+import shutil
+import tempfile
+import zipfile
+
+import click
+import click_odoo
+from _dbutils import db_exists
+from click_odoo import odoo
+
+
+def _dump_db(dbname, backup):
+    with tempfile.NamedTemporaryFile() as f:
+        cmd = ["pg_dump", "--no-owner", "--format=c", dbname, "--file=" + f.name]
+        odoo.tools.exec_pg_command(*cmd)
+        backup.addfile(f.name, "dump.sql")
+
+
+def _create_manifest(cr, dbname, backup):
+    with tempfile.NamedTemporaryFile() as f:
+        json.dump(odoo.service.db.dump_db_manifest(cr), f, indent=4)
+        f.seek(0)
+        backup.addfile(f.name, "manifest.json")
+
+
+def _backup_filestore(dbname, backup):
+    filestore_source = odoo.tools.config.filestore(dbname)
+    if os.path.isdir(filestore_source):
+        backup.copytree(filestore_source)
+
+
+class _Backup(object):
+    def __init__(self, path, format):
+        assert format in ("zip", "folder"), "Support only 'zip' or 'folder'"
+        self._zipFile = None
+        self._path = path
+        self.format = format
+
+    def __enter__(self):
+        if self.format == "zip":
+            self._zipFile = zipfile.ZipFile(
+                self._path, "w", compression=zipfile.ZIP_DEFLATED, allowZip64=True
+            )
+        else:
+            os.mkdir(self._path)
+        return self
+
+    def __exit__(self, *arg):
+        if self._zipFile:
+            self._zipFile.close()
+
+    def copytree(self, src):
+        """Recursively copy a directory tree to tje backup.
+        Dirname of src dir is the root of copied files into the archive
+        """
+        if self._zipFile:
+            len_prefix = len(os.path.dirname(src)) + 1
+            for dirpath, _dirnames, filenames in os.walk(src):
+                for fname in filenames:
+                    path = os.path.normpath(os.path.join(dirpath, fname))
+                    if os.path.isfile(path):
+                        self._zipFile.write(path, path[len_prefix:])
+        else:
+            dest = os.path.join(self._path, os.path.basename(src))
+            shutil.copytree(src, dest)
+
+    def addfile(self, filename, arcname):
+        """Add a file to the backup.
+
+        : param filename : the path to the souce file
+        : arcname : the path into the backup
+        """
+        if self._zipFile:
+            self._zipFile.write(filename, arcname)
+        else:
+            shutil.copyfile(filename, os.path.join(self._path, arcname))
+
+
+@click.command()
+@click_odoo.env_options(
+    default_log_level="warn", with_database=False, with_rollback=False
+)
+@click.option(
+    "--unless-dest-exists",
+    is_flag=True,
+    help="Don't report error if destination file/folder already exists.",
+)
```

Comment:
Since this overwrites the target backup files, I would name this `--force` instead, because `--unless-exists` is a noop in other scripts where it is used.